### PR TITLE
User pragmas hide effect specs bug fix. Fixes #7216

### LINKED ISSUE #7216: User pragmas hide effect specifications from sempass2
The following compiles while it shouldn't:

```
{.pragma: my_pragma, raises: [].}

proc test1 {.my_pragma.} =
  raise newException(Exception, "msg")
```

### PR DESCRIPTION
**Pull request by cooldome**
Fix for  #7216.
Test included

### CODE REVIEW

File: tests/pragmas/tuserpragma.nim [hunk 1]
```diff
@@ -1,7 +1,19 @@
...
 proc myproc(x, y: int): int {.rtl} =
-  nil
+  discard
 
+#####################
 
+# bug #7216
+{.pragma: my_pragma, raises: [].}
+
+doAssert(not compiles(
```
Comment (Araq) [inline tests/pragmas/tuserpragma.nim]: This is not really what ``compiles`` is about. The test case should also check the compiler's error message too.

Comment (cooldome) [reply tests/pragmas/tuserpragma.nim]: Should I put in the separate test file then?

Comment (Araq) [reply tests/pragmas/tuserpragma.nim]: Tests can have ``errormsg`` specs.

Comment (Araq) [reply tests/pragmas/tuserpragma.nim]: Oh I see what mean. Yes, separate test :-(

**Comment by cooldome** [comment]
I have grepped the code by whichPragma and the problem is more systematic than I thought. I suggest I write a pragma iterator that will walk pragmas without skipping pragmas in pragmas and return (index, TSpecialKeyword) and port everything to it. How does it sound?

**Comment by cooldome** [comment]
Another option is to properly linearise pragmas during pragma sem checks, semcheck code will have to adapt to the fact number of pragmas can increase/decrease as you process them

**Comment by cooldome** [comment]
After a more close look at the code I realized that almost noone bothers to traverse pragmas recursively. Effect specs is definitely not the only affected. More radical approach is required. 
Solution: linearize pragmas such that simple for loop is now safe to apply after semcheck

File: compiler/pragmas.nim [hunk 2]
```diff
@@ -691,14 +691,18 @@ proc singlePragma(c: PContext, sym: PSym, n: PNode, i: int,
...
+    if userPragma.ast.len == 0:
+      delete(n.sons, i)
+      dec i
+    else:
+      inc c.instCounter
+      if c.instCounter > 100:
+        globalError(it.info, errRecursiveDependencyX, userPragma.name.s)
+      
+      pragma(c, sym, userPragma.ast, validPragmas)
+      n.sons[i..i] = userPragma.ast.sons
```
Comment (Araq) [inline compiler/pragmas.nim]: I don't like these kind of AST transformations. Hard to review.

File: compiler/pragmas.nim [hunk 3]
```diff
@@ -691,14 +691,18 @@ proc singlePragma(c: PContext, sym: PSym, n: PNode, i: int,
...
-    inc c.instCounter
-    if c.instCounter > 100:
-      globalError(it.info, errRecursiveDependencyX, userPragma.name.s)
-    pragma(c, sym, userPragma.ast, validPragmas)
-    # ensure the pragma is also remember for generic instantiations in other
-    # modules:
-    n.sons[i] = userPragma.ast
-    dec c.instCounter
+    if userPragma.ast.len == 0:
+      delete(n.sons, i)
```
Comment (Araq) [inline compiler/pragmas.nim]: Why?

File: compiler/pragmas.nim (hunk 2 above)
Comment (cooldome) [reply compiler/pragmas.nim]: Agree they are not straight forward, but I don't have better ideas. Let me know if you have some.
I will add comments to the code to make easier to follow.

File: compiler/pragmas.nim (hunk 3 above)
Comment (cooldome) [reply compiler/pragmas.nim]: If user pragma is ended-up being empty (not uncommon situation) after expansion of user pragma you will get one pragma less: 
Before expansion: ``proc x {.my_pragma, inline.}``
After expansion: ``proc x {.inline.}``

P.S.
I have found a way to merge `` n.sons[i..i] = userPragma.ast.sons`` and ``delete(n.sons, i)`` branches into one. I will submit an updated version

**Comment by cooldome** [comment]
Slightly easier to follow version is submitted. 
processPragma was changed such that ``userPragma.ast`` is empty seq when user pragma is empty, previously it was ``nil``. This was causing issues for line ``n.sons[i..i] = userPragma.ast.sons`` in ``singlePragma`` for empty user pragmas.

**Comment by cooldome** [comment]
test failures are unrelated to the change

**Pull request merged**